Implement basic frame class

nealian opened the pull request:
closes #4 

Potentially, anyway

nealian committed:
Implement basic frame class

nealian committed:
Add raw() frame getter, to get raw frame "bytes" (as strings).

0culus commented on src/Frame.h:
```diff
@@ -0,0 +1,35 @@
+#ifndef __FRAME_H__
```
``` c++
#pragma once
```

has the same effect as include guards and looks better.

0culus commented on src/Frame.h:
```diff
@@ -0,0 +1,35 @@
+
+class Frame {
+public:
+    // Getters
```
Is there a need to modify a `Frame` object at runtime? If not, this could be done more cleanly in the ctor with initializer lists and not using a plethora of getters and setters.

nealian replied:
Could you expand on this?

0culus replied:
Like, if you don't specifically need to get and set, you can set up your constructor something like this (my node class from the last project):

``` c++
struct Node {
  std::string       _name;
  unsigned short    _port, _port_extra;
  std::stringstream _config_data;
  //TCPSocket         _sock, _sock_extra;
  bool              _terminate = false;
  std::queue<Message> _to_send;

  Node(std::string name,
       unsigned short port,
       std::stringstream data) : _name{ name },
                         _port{ port },
                         _config_data{ std::move(data) } { }
  Node(std::string name,
       unsigned short port,
       unsigned short port_extra,
       std::stringstream data) : _name{ name },
                               _port{ port },
                               _port_extra{ port_extra },
                               _config_data{ std::move(data) } { }
  Node(const Node&) = default;
  Node(Node&&)      = default;

  void listen_and_print();
  std::thread lp_thread() {
    return std::thread([this] { listen_and_print(); });
  }
  void connect_and_send();
  std::thread cs_thread() {
    return std::thread([this] { connect_and_send(); });
  }
  void handle_client(std::unique_ptr<TCPSocket> sock);
};
```

Note that in C++, structs are the same as classes except all members default to public.

nealian replied:
Fair 'nuff

0culus replied:
I think it will simplify things...especially since he wants things threaded again. All those getters and setters will have to be made thread safe.

0culus commented on src/Frame.h:
```diff
@@ -0,0 +1,35 @@
+#include <string>
+#include <cstdint>
+
+using namespace std;
```
I prefer to not  `using namespace std`; Prefixing stuff in the std namespace with `std::` helps avoid any confusion with namespaces we may define.

nealian replied:
That's fair; mostly I was copying the style of my code from last year, haha

0culus commented on src/Frame.h:
```diff
@@ -0,0 +1,35 @@
+    // Constructors
+    Frame();
+    Frame(string contents);
+    Frame(uint8_t src, uint8_t dst, string data);
```
Might be a good idea to define copy and move ctors too:

``` c++
Frame(const Frame&) = default;
Frame(Frame&&) = default;
```

0culus replied:
Or alternatively, explicitly delete them if we decide we don't want this object to either be copyable, moveable, or both.

nealian replied:
I honestly like the second; since we're not likely to be modifying these (except perhaps for acknowledgement, in which case we have likely done what we need to do anyway), there is no real reason to generate a copy in the first place.  Similar logic for the move case.

0culus replied:
Sounds good.

nealian committed:
Frame class: modifications round one
Cleaner header guards
Be explicit with namespace
Remove unnecessary setters
Explicitly delete the copy and move constructors
cstdint is C++11

nealian committed:
Frame class: modifications round two
Remove append setter (I guess I missed it?)
Remove empty constructor
Simplify the simple-case constructor
